DOC: improve development debugging doc (#16083)
numpy/numpy: doc/source/dev/development_environment.rst

---------

Another frequently asked question is "How do I debug C code inside NumPy?".
First you must obtain the correct version of gdb (often the default on

> Linux), which requires an embedded Python. To see which version of Python

Reviewer: Comparing this to the deleted text (in what used to be lines 231-232 in the diff) it seems that the meaning has changed a bit. For example, the parenthetical statement below was referring to the fact that the Python extensions for gdb are often installed by default on Linux, whereas it's placement here indicates that "the correct version" of gdb is the default on Linux.

I like the addition of the example below explaining how to test the embedded Python from within the gdb session, but I think that the wording here needs a little work to recapture the original meaning of the deleted sentence.

Author: yes, the meaning here is unclear, added more accurate version description.

> Linux), which requires the embedded Python interpreter version higher than 3.6.

Author: Googled it, 'Python version' is used more frequently than 'Python interpreter version'

> Linux), which requires the embedded Python interpreter version higher than 3.6.

Reviewer: Good to know :)... but my point was more about whether it was necessary to include the version information at all, i.e. eliminate the "which requires the embedded Python interpreter version higher than 3.6". This suggestion is predicated on the fact that `gdb` is not sensitive to the version of Python that is running - if that's not true (i.e. gdb doesn't work with Python < 3.6) then this statement can be left in.

> Linux), which requires the embedded Python interpreter version higher than 3.6.

Author: well, in my practice, debug C code inside NumPy requires gdb with python > 3.6, other versions, such as gdb that not support Python scripting or gdb with python version < 3.6, is failed to debug.

> To see which version of Python is embedded inside your gdb, do the following::

Reviewer 2: I am able to run gdb both with 3.5 and 3.8. I think we should avoid mentioning python version here.

Author: I tried python2.7 and failed to debug, should I just mention python3 without minor version?

Reviewer: I agree with @anirudh2290  - the Python version embedded in gdb feels like a gdb-configuration-specific detail that we don't need to get entrenched on. IMO it's easiest to simply leave the version info out.

If users end up having trouble with a gdb-Python interaction, they'll be looking to docs other than NumPy's to solve it anyways.

Author: ok then, specific python version removed.

> First you must obtain the version of gdb higher than 7.0 (which includes an
> embedded Python), secondly make sure the embedded Python is 3.x. To see
> which version of Python is embedded inside your gdb, do the following::

Reviewer: ```suggestion
First, ensure that you have gdb installed on your system with the Python
extensions (often the default on Linux). You can see which version of
Python is running inside gdb to verify your setup::
```
I still don't think the proposed wording captures the original meaning very well. Here's a suggestion for wording that I think captures the original meaning, avoids any potential pitfalls related to gdb or Python version info, and connects to the very nice example that you've included for verifying that gdb is configured appropriately with the Python extensions.

LMK what you think of this wording!

Author: Thanks for taking the trouble to help me，I'm struggling at import the python version verifying example smoothly without mention any gdb or python version,  did not find a way to do that(so I mention them all). Your statement is more precise and fluent.

The easiest way to do this is to first write a Python script that invokes the C
code whose execution you want to debug. For instance ``mytest.py``::

    import numpy as np
    x = np.arange(5)
    np.empty_like(x)

> The next thing to do is to obtain the correct version of for gdb(often the default

Reviewer: ```suggestion
The next thing to do is to obtain the correct version of for gdb (often the default
```

Reviewer 2: nit: also `of for` -> `of`.

> The next thing to do is to obtain the correct version of for gdb(often the default

Reviewer: It might make sense to include the instructions on checking that you have a debugger installed on the system *before* the above sentence. For example:

Another frequently asked question is "How do I debug C code inside NumPy?".
First you must ensure you have a debugger installed... <rest of the explanation>.

Author: you are right, environment should configured before writing a debugger file.

    (gdb) run

The execution will now stop at the corresponding C function and you can step
through it as usual. A number of useful Python-specific commands are available.
For example to see where in the Python code you are, use ``py-list``.  For more

> details, see `DebuggingWithGdb`_. Here lists the most commonly used commands::

Reviewer: ```suggestion
details, see `DebuggingWithGdb`_. Here are some commonly used commands::
```

>     list- List specified function or line

Reviewer: I haven't checked how this is rendered, but perhaps a bulleted-list would be more appropriate than a literal block? It might be fine.

Author: @rossbar do you have a bulleted-list example?
